Fix deployment schedule

fregante opened the pull request:
As mentioned in #25, this is now possible: one workflow, only deploys when necessary.

fregante committed:
Delete deploy-automatic.yml

fregante committed:
Merge schedule into main workflow

fregante commented on .github/workflowsdeployment.yml:
```diff
@@ -11,7 +13,27 @@ on:
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+      - name: Create tag if necessary
+        if: github.event_name == 'schedule'
```
Technically this should be on the whole job, but then `Build` won't run at all since it `needs` this job.

fregante committed:
Lint

fregante commented on .github/workflowsdeployment.yml:
```diff
@@ -11,7 +13,27 @@ on:
+      - name: Create tag if necessary
+        if: github.event_name == 'schedule'
+        run: |
+          git fetch --tags
+          if [[ $(git tag -l --points-at HEAD) = '' ]]; then
+            echo '::set-output name=created::yes'
+            export VER=$(npx daily-version)
+            git tag $VER -m $VER
+            git push origin $VER
+          fi
```
I will probably wrap this into an action, e.g. `fregante/daily-version`

fregante committed:
Fix name

notlmn commented:
Sorry @fregante, my internship has been killing me lately, especially with the burnout after that.

Renamed `GH_TOKEN` to `TAGGER_TOKEN` (as asked in https://github.com/notlmn/browser-extension-template/issues/25#issuecomment-608349740).

Feel free to use this repo to test any possible cases you have, even when I'm offline. I think that is what this repo is intended for.

We should test with changes similar to #26, with maybe also changes to hours and minutes in cron (to every 5 minutes for testing?) and update that later, once we get to a working solution.

Thanks for bearing with me!

fregante commented:
No worries, that’s no longer needed and you don’t need to do anything except maybe review once I’m done. I just need to publish an action and this should be good to go

notlmn commented on .github/workflows/deployment.yml:
```diff
@@ -11,7 +13,27 @@ on:
 jobs:
 
-  build:
+  Version:
```
```suggestion
  version:
```

notlmn commented on .github/workflows/deployment.yml:
```diff
@@ -27,8 +49,8 @@ jobs:
 
-  deploy-to-chrome:
-    needs: build
+  For Chrome:
+    needs: Build
```
```suggestion
  release-chrome:
    needs: build
```

notlmn commented on .github/workflows/deployment.yml:
```diff
@@ -42,8 +64,8 @@ jobs:
 
-  deploy-to-firefox:
-    needs: build
+  For Firefox:
+    needs: Build
```
```suggestion
  release-firefox:
    needs: build
```

notlmn commented on .github/workflows/deployment.yml:
```diff
@@ -11,7 +13,27 @@ on:
+            git push origin $VER
+          fi
+
+  Build:
+    needs: Version
+    if: github.event_name == 'push' || needs.Version.outputs.created
```
```suggestion
  build:
    needs: version
    if: github.event_name == 'push' || needs.version.outputs.created
```

notlmn reviewed (dismissed):
Apparently GHA job names cannot contain spaces.

Refer https://github.com/notlmn/copy-as-markdown/actions/runs/101354661 for error message.

fregante commented:
I've been capitalizing them because the names appear in the UI. Seems like an easy way to make them look better (without using an extra "name" property)

<img width="380" alt="" src="https://user-images.githubusercontent.com/1402241/81560036-e54fb000-9390-11ea-87e7-a8db37fd2b52.png">


I can remove the `For ` so it will look like

- Deployment / Version
- Deployment / Build
- Deployment / Chrome
- Deployment / Firefox

notlmn commented:
🎉 This works https://github.com/notlmn/copy-as-markdown/runs/684846163

notlmn commented on .github/workflows/deployment.yml:
```diff
@@ -11,7 +13,27 @@ on:
+        if: github.event_name == 'schedule'
+        run: |
+          git fetch --tags
+          if [[ $(git tag -l --points-at HEAD) = '' ]]; then
+            echo '::set-output name=created::yes'
+            export VER=$(npx daily-version)
+            git tag $VER -m $VER
+            git push origin $VER
+          fi
```
Maybe we should add an `else` clause here and `echo` that there are no new changes to publish, hence skipping.

Can do the same if we have successfully created a tag.

fregante replied:
Sounds good. I hope to find the time to wrap this into an action so we reduce the amount of code to copy/maintain in all repos.

fregante commented:
I don't think this parts work correctly yet though:

```yml
    - name: Update extension’s meta
      run: |
        export VER=$(echo "$GITHUB_REF" | sed -e 's/refs\/tags\///')
        echo https://github.com/$GITHUB_REPOSITORY/tree/$VER > distribution/SOURCE_URL
        npm run version
```

`$GITHUB_REF` won't be a tag when it's a scheduled run; I think this will produce:

```
https://github.com/notlmn/browser-extension-template/tree//refs/heads/master
```

The version should be either parsed from $GITHUB_REF (if a workflow is launched from a tag, it should match **that** tag) or "the latest tag" if missing.

Suggestions?

notlmn commented:
> The version should be either parsed from $GITHUB_REF (if a workflow is launched from a tag, it should match **that** tag) or "the latest tag" if missing.

Either ways, shouldn't we always pick the latest tag available?

``` diff
    - name: Update extension’s meta
      run: |
-       export VER=$(echo "$GITHUB_REF" | sed -e 's/refs\/tags\///')
+       export VER=$(git describe --tags --abbrev=0)
        echo https://github.com/$GITHUB_REPOSITORY/tree/$VER > distribution/SOURCE_URL
        npm run version
```

Thoughts?

fregante commented:
> shouldn't we always pick the latest tag available?

No 👇 
> if a workflow is launched from a tag, it should match **that** tag

Meaning: if a user somehow triggers the workflow by pushing an older tag, **that tag** should be deployed, not the latest on master.

fregante commented:
However, if `git describe --tags --abbrev=0` picks **that tag** then it's great!

notlmn commented:
> However, if `git describe --tags --abbrev=0` picks **that tag** then it's great!

I think it would not.

--------

We could add a condition then, if the job is run because of a schedule, let's pick version from `$GITHUB_REF` and use `git describe --tags --abbrev=0` otherwise.

This check could be inside the bash script (compact), or add another conditional step (which I don't like too).

Again, thoughts?

fregante committed:
Use daily-version-action

fregante committed:
Fix "Update extension’s meta" step

fregante commented:
I think that thanks to `daily-version-action` we no longer need to parse the tag nor add further logic here; we already have a `version` output.

Note: none of this is tested yet, including `daily-version-action` itself.

fregante commented:
Nope. Hmmm something needs to change.

- Versions can only be created by `daily-version-action` on the latest commit, whether it's a push to master or a scheduled run.

The current condition on `daily-version-action` means that `needs.Version.outputs.version` will only be set on `schedule`

This affects both our deployment logic and `daily-version-action` itself since the output can no longer be relied upon (like I thought in https://github.com/notlmn/browser-extension-template/pull/28/commits/c4738418da2761971f849feebeac204b379d49d2)

Perhaps a solution would be to move the condition inside the action so the action will always set `version`.

```yml
      - name: Create tag if necessary
        uses: fregante/daily-version-action@v1
        with:
          # This `if` will be mandatory to avoid mistakes
          if: github.event_name == 'schedule'
```

Better yet:


```yml
      - name: Create tag if necessary
        uses: fregante/daily-version-action@v1
        with:
          # At least one condition
          schedule: true # Only on cronjobs
          push: master # Only when it's the latest on master
```

notlmn commented:
> Perhaps a solution would be to move the condition inside the action so the action will always set `version`.
> ...

Excellent. As you've mentioned we don't have to move conditions inside the action, but have options passed into it.

-------

> Better yet:
> ...

I'd suggest a more generic approach.

``` yaml
      - name: Create tag if necessary
        uses: fregante/daily-version-action@v1
        with:
          events:
            - schedule
            - push
            - ...
          branches:
            - master
            - dev
```

Sorry I'm bad at YAML, but you get the idea.

fregante commented:
Maybe `daily-version-action` doesn't need the conditions at all since the whole workflow should be dedicated to **deployment**:

```yml
on:
  schedule:
    - cron: '31 13 * * 3'
```
```yml
on:
  push:
    tags:
      - "20.*"
      - "21.*"
      - "22.*"
      - "23.*"
      - "24.*"
```
```yml
on:
  push:
    branch:
      - master
```

Or a combination of `schedule` and `tag` like in our case.

---

I only need to exclude the tag creation, automatically, when the event is already a tag.

fregante committed:
Update for next daily-version-action version

- Cronjob is already explained by tooltip
- No condition: https://github.com/fregante/daily-version-action/pull/4
- Not git setup: https://github.com/fregante/daily-version-action/issues/2

fregante committed:
Fix Chrome/Firefox jobs names

fregante committed:
Jobs must declare outputs

fregante commented on .github/workflows/deployment.yml:
```diff
@@ -11,24 +13,38 @@ on:
-  build:
+  Version:
+    runs-on: ubuntu-latest
+    outputs:
+      version: ${{ steps.daily-version.outputs.version }}
+      created: ${{ steps.daily-version.outputs.created }}
+    steps:
+      - uses: actions/checkout@v2
+      - uses: fregante/daily-version-action@v1
+        name: Create tag if necessary
+        id: daily-version
```
1. Outputs must be declared
1. `fregante/daily-version-action` was tested on:
	- push
		- commit
			> HEAD isn’t tagged. `daily-version-action` will create `20.6.6.1327`
		- tag
			> Run triggered by tag `20.6.6.1321`. No new tags will be created by `daily-version-action`.
	- any non-push event:
		- needs tag
			> HEAD isn’t tagged. `daily-version-action` will create `20.6.6.1326`
		- doesn't need tag
			> No new commits since the last tag (20.6.6.1324). No new tags will be created by `daily-version-action`.

fregante commented on .github/workflows/deployment.yml:
```diff
@@ -11,24 +13,38 @@ on:
+
+  Build:
+    needs: Version
+    if: github.event_name == 'push' || needs.Version.outputs.created
```
Correctly passes all tested situations:
- any `push` has a tag, either because the push itself is tag or it was created by the `Version` step
- the `created` output is correctly set only if HEAD wasn't initially tagged, in any situation

fregante commented on .github/workflows/deployment.yml:
```diff
@@ -11,24 +13,38 @@ on:
+        VER: ${{ needs.Version.outputs.version }}
       run: |
-        export VER=$(echo "$GITHUB_REF" | sed -e 's/refs\/tags\///')
         echo https://github.com/$GITHUB_REPOSITORY/tree/$VER > distribution/SOURCE_URL
```
Tested: $VER is correctly set here

fregante reviewed (commented):
✅ I think we're good to go!

fregante commented:
Rebased on top of #24

fregante commented:
For the record, I tried merging the Chrome/Firefox steps with a `matrix`, but the result is surprisingly not shorter/cleaner 🤷‍♂️, so it's not worth it

```gql
  Publish:
    name: ${{ matrix.name }}
    needs: Build
    runs-on: ubuntu-latest
    
    strategy:
      matrix:
        include:
          - name: Chrome
            command: cd artifact && npx chrome-webstore-upload-cli@1 upload --auto-publish
          - name: Firefox
            command: cd artifact && npx web-ext-submit@4
    steps:
    - uses: actions/download-artifact@v2
    - run: ${{ matrix.command }}
      env:
        # Chrome
        EXTENSION_ID: ${{ secrets.EXTENSION_ID }}
        CLIENT_ID: ${{ secrets.CLIENT_ID }}
        CLIENT_SECRET: ${{ secrets.CLIENT_SECRET }}
        REFRESH_TOKEN: ${{ secrets.REFRESH_TOKEN }}
        # Firefox
        WEB_EXT_API_KEY: ${{ secrets.WEB_EXT_API_KEY }}
        WEB_EXT_API_SECRET: ${{ secrets.WEB_EXT_API_SECRET }}
```

notlmn reviewed (approved):
LGTM!

fregante commented:
Let's merge this to ensure it works correctly. Then we can add it to the other extensions

fregante commented:
It worked!

- Schedule run: https://github.com/notlmn/browser-extension-template/actions/runs/131034468
- Tag generated: https://github.com/notlmn/browser-extension-template/releases/tag/20.6.10
- SOURCE_URL contains `https://github.com/notlmn/browser-extension-template/tree/20.6.10`

Last thing that we hope works correctly is the "HEAD is tagged, skipping release" behavior. For that, we'll have to wait a week without commits.

fregante commented:
> Last thing that we hope works correctly is the "HEAD is tagged, skipping release" behavior. For that, we'll have to wait a week without commits.

Confirmed!

<img width="786" alt="" src="https://user-images.githubusercontent.com/1402241/85019447-26e13100-b16f-11ea-956e-186640e5d093.png">
